### ISSUE TITLE
dgCMatrix crashes fit_generator

### ISSUE THREAD

**Issue by alanault** [rstudio/keras#187]
Hi there,

Really loving the package - had an issue processing sparse matrices output from the text2vec package for text classification. The dataset is too large to train in memory, so I've been trying to use the fit_generator approach as per the documentation.

If I pass a sparse matrix as an x argument the fit_generator function prints the initial batch to console and then crashes. 

I also wondered the best approach for providing a validation set using this approach? Can this be passed from the sampling generator, or should there be a separate set, which must fit in memory?

We can reproduce the crash using the iris dataset:

```
library(keras)

# Working example-----------------------------------------
# define model
model <- keras::keras_model_sequential()

# create simple network structure
model %>%
  keras::layer_dense(units = 16, input_shape = 4, activation = "relu" ) %>%
  keras::layer_dense(units = 4, activation = 'softmax')

# compile
model %>%
  keras::compile(optimizer = "rmsprop",
                 loss = 'categorical_crossentropy',
                 metrics = c('accuracy'))

# Make a sample generator to pass the data in batches
# Rows is a matrix as we're one hot encoding
sampling_generator <- function(X_data, Y_data, batch_size) {
  function() {
    rows <- sample(1:nrow(X_data), batch_size, replace = TRUE)
    list(X_data[rows,], Y_data[rows,])
  }
}

# fit the model using the sampling generator
model %>%
  fit_generator(sampling_generator(as.matrix(iris [ , 1:4]),
                                   to_categorical(iris$Species),
                                   batch_size = 5),
                steps_per_epoch = nrow(iris) / 5,
                epochs = 50)

# Version which crashes-----------------------------------

# turn iris inputs into a sparse matrix
library(Matrix)
iris_sparse <- as(as.matrix(iris [, 1:4]), "dgCMatrix")

# try fit generator again and crashes
model %>%
  fit_generator(sampling_generator(iris_sparse,
                                   to_categorical(iris$Species),
                                   batch_size = 5),
                steps_per_epoch = nrow(iris) / 5,
                epochs = 50)
```

Any suggestions?

**Comment by jjallaire** [comment]
Interesting. I'm not able to reproduce the crash with this environment (which includes the current dev master of the keras package):

```
> sessionInfo()
R version 3.4.1 (2017-06-30)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS Sierra 10.12.6

Matrix products: default
BLAS: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] Matrix_1.2-10    keras_2.0.8.9010

loaded via a namespace (and not attached):
 [1] compiler_3.4.1        magrittr_1.5          R6_2.2.2              tools_3.4.1          
 [5] whisker_0.3-2         base64enc_0.1-3       yaml_2.1.14           Rcpp_0.12.13         
 [9] reticulate_1.2.0.9012 grid_3.4.1            tensorflow_1.4.0.9007 zeallot_0.0.6        
[13] jsonlite_1.5          lattice_0.20-35       tfruns_1.0        
```

What platform are you running on and what versions of packages are you running?

**Comment by alanault** [comment]
Curious - we're on almost identical systems, other than small version increments.

I'll work through each of the packages installing the latest minor versions from github and see if it persists when we're on identical packages 

```
R version 3.4.1 (2017-06-30)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS Sierra 10.12.6

Matrix products: default
BLAS: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRlapack.dylib

locale:
[1] en_GB.UTF-8/en_GB.UTF-8/en_GB.UTF-8/C/en_GB.UTF-8/en_GB.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] Matrix_1.2-10   keras_2.0.8     devtools_1.13.3

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.12      lattice_0.20-35   digest_0.6.12     withr_2.0.0       grid_3.4.1       
 [6] R6_2.2.2          jsonlite_1.5      magrittr_1.5      tfruns_0.9.1.9002 whisker_0.3-2    
[11] reticulate_1.2    tools_3.4.1       yaml_2.1.14       compiler_3.4.1    base64enc_0.1-3  
[16] tensorflow_1.4    memoise_1.1.0    
```

**Comment by alanault** [comment]
Updated to keras_2.0.8.9010 and reticulate to reticulate_1.2.0.9012 and we're now working!

Thanks for the help!

Any thoughts on validation? Do I have to commit to a validation set up front, or can I also pass validation samples through to the fit_generator?

Many thanks
Alan

**Comment by jjallaire** [comment]
You can pass a generator in the `validation_data` argument, like this:

```r
history <- model %>% fit_generator(
  train_gen,
  steps_per_epoch = 500,
  epochs = 40,
  validation_data = val_gen,
  validation_steps = val_steps
)
```

**Comment by alanault** [comment]
That works just fine - thanks very much!
